<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Assignment08 · BISC195.jl</title><link rel="canonical" href="https://wellesley-bisc195.github.io/BISC195.jl/Assignments/Assignment08/"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script><link href="../../assets/admonitions.css" rel="stylesheet" type="text/css"/></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img class="docs-light-only" src="../../assets/logo.svg" alt="BISC195.jl logo"/><img class="docs-dark-only" src="../../assets/logo-dark.svg" alt="BISC195.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">BISC195.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../../Syllabus/">Syllabus</a></li><li><span class="tocitem">Lesson 1 - Getting Started</span><ul><li><a class="tocitem" href="../../Lessons/Lesson1/">Lesson docs</a></li><li><a class="tocitem" href="../Assignment01/">Assignment01</a></li></ul></li><li><span class="tocitem">Lesson 2 - Data Types and Functions</span><ul><li><a class="tocitem" href="../../Lessons/Lesson2/">Lesson docs</a></li><li><a class="tocitem" href="../Assignment02/">Assignment02</a></li></ul></li><li><span class="tocitem">Lesson 3 - Building functions and composition</span><ul><li><a class="tocitem" href="../../Lessons/Lesson3/">Lesson docs</a></li><li><a class="tocitem" href="../Assignment03/">Assignment03</a></li></ul></li><li><span class="tocitem">Lesson 4 - Building functions and composition</span><ul><li><a class="tocitem" href="../../Lessons/Lesson4/">Lesson docs</a></li><li><a class="tocitem" href="../Assignment04/">Assignment04</a></li></ul></li><li><span class="tocitem">Lesson 5 - Arrays</span><ul><li><a class="tocitem" href="../../Lessons/Lesson5/">Lesson docs</a></li><li><a class="tocitem" href="../Assignment05/">Assignment05</a></li></ul></li><li><span class="tocitem">Lesson 6 - Miscellany</span><ul><li><a class="tocitem" href="../../Lessons/Lesson6/">Lesson docs</a></li><li><a class="tocitem" href="../Assignment06/">Assignment06</a></li></ul></li><li><span class="tocitem">Lesson 7 - Code Re-Use: Packages and Environments</span><ul><li><a class="tocitem" href="../../Lessons/Lesson7/">Lesson docs</a></li><li><a class="tocitem" href="../Assignment07/">Assignment07</a></li></ul></li><li><span class="tocitem">Lesson 8 - Code Re-Use: Packages and Environments</span><ul><li><a class="tocitem" href="../../Lessons/Lesson8/">Lesson docs</a></li><li class="is-active"><a class="tocitem" href>Assignment08</a><ul class="internal"><li><a class="tocitem" href="#Check-list"><span>Check list</span></a></li><li><a class="tocitem" href="#Wrapping-up-Assignment07-Package"><span>Wrapping up Assignment07 Package</span></a></li><li><a class="tocitem" href="#Building-an-analysis-repository"><span>Building an analysis repository</span></a></li><li><a class="tocitem" href="#Getting-the-data"><span>Getting the data</span></a></li><li><a class="tocitem" href="#Setting-up-julia-project"><span>Setting up julia project</span></a></li><li><a class="tocitem" href="#Using-the-notebook"><span>Using the notebook</span></a></li><li><a class="tocitem" href="#No-tests"><span>No tests</span></a></li></ul></li></ul></li><li><a class="tocitem" href="../Assignments/">Assignment Index</a></li><li><a class="tocitem" href="../../Lectures/Lectures/">Lecture Index</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Lesson 8 - Code Re-Use: Packages and Environments</a></li><li class="is-active"><a href>Assignment08</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Assignment08</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/wellesley-bisc195/BISC195.jl/blob/master/docs/src/Assignments/Assignment08.md#L" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="assignment08"><a class="docs-heading-anchor" href="#assignment08">Assignment 08</a><a id="assignment08-1"></a><a class="docs-heading-anchor-permalink" href="#assignment08" title="Permalink"></a></h1><p><a href="#assignment08"><img src="https://img.shields.io/badge/Due-7%2F17%2F2020-orange?style=for-the-badge&amp;logo=open%20badges" alt="Assignment 08 - Due"/></a></p><h2 id="Check-list"><a class="docs-heading-anchor" href="#Check-list">Check list</a><a id="Check-list-1"></a><a class="docs-heading-anchor-permalink" href="#Check-list" title="Permalink"></a></h2><p>There are a lot of different tasks introduced in this assignment, and you will not have automated tests to tell you when you&#39;re done. Here&#39;s a checklist to help you keep track, which Kevin will add to your github repo once you&#39;ve created it and shared it with him.</p><ul><li>[ ] Code repo (from Assignment 7), renamed and in public repo<ul><li>[ ] tests still passing</li><li>[ ] kmer counting function, with tests added</li></ul></li><li>[ ] Analysis repo complete<ul><li>[ ] README has heading 1 title and bold author line</li><li>[ ] <code>Project.toml</code> with code repo, <code>Revise</code>, <code>DataFrames</code>, and <code>Plots</code> as dependencies</li><li>[ ] <code>data/</code> directory with explanation for how to download data<ul><li>[ ] <code>cov-sequences.fasta</code></li></ul></li><li>[ ] <code>notebooks/</code> directory with <code>1-sequence-stats.md</code><ul><li>[ ] Including code and description for calculating mean and standard deviations     for sequence length and gc content</li></ul></li></ul></li></ul><h2 id="Wrapping-up-Assignment07-Package"><a class="docs-heading-anchor" href="#Wrapping-up-Assignment07-Package">Wrapping up Assignment07 Package</a><a id="Wrapping-up-Assignment07-Package-1"></a><a class="docs-heading-anchor-permalink" href="#Wrapping-up-Assignment07-Package" title="Permalink"></a></h2><p>In the <a href="../Assignment07/#assignment07">previous assignment</a>, we created a julia package that you&#39;re going to be re-using for the rest of the course. But it doesn&#39;t have a very good name, and at the moment, your assignments are in private repos,  which makes some stuff (like using it for automated tests) harder.</p><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>Before continuing, make sure that you&#39;ve successfully completed assignment 7 and all of the tests are passing</p></div></div><p>First, let&#39;s make a new, public home for your bioinformatics package.</p><ol><li><p>On github, create a new repository:</p><p><img src="../assets/img/new-repo.png" alt="Create repo"/></p></li><li><p>Give it a good name - Julia package names usually end with <code>.jl</code>. Be sure that the repo is set to <code>public</code>. <img src="../assets/img/repo-name.png" alt="Create repo"/></p></li><li><p>Copy the url of the repository</p><p><img src="../assets/img/clone-repo.png" alt="Create repo"/></p></li><li><p>On your computer, navigate to your <code>Assignment07-&lt;username&gt;</code> directory. We are now going to point <code>git</code> to this new location, so that when you make new changes in the future, they will go there instead.</p><p>In the terminal,</p><pre><code class="language-sh">$ git remote set-url origin &lt;paste URL here&gt;
$ git push -u origin trunk
Enumerating objects: 61, done.
Counting objects: 100% (61/61), done.
Delta compression using up to 12 threads
Compressing objects: 100% (46/46), done.
Writing objects: 100% (61/61), 76.07 KiB | 1.46 MiB/s, done.
Total 61 (delta 21), reused 24 (delta 7), pack-reused 0
remote: Resolving deltas: 100% (21/21), done.
To github.com:kescobo/BioinformaticsBISC195.jl.git
 * [new branch]      trunk -&gt; trunk
Branch &#39;trunk&#39; set up to track remote branch &#39;trunk&#39; from &#39;origin&#39;.</code></pre></li><li><p>Now let&#39;s rename the module in the julia code. Open the assignment <em>directory</em> in VS Code. There are a few places that the module name currently exists - use <code>cmd+shift+F</code> (<code>ctrl+shift+F</code> on Windows) to open the search panel (you can also click the magnifying glass icon in the sidebar) and search for <code>Assignment07</code>. </p><p>Replace <code>Assignment07</code> with whatever you named your module, and then rename the main module file at <code>src/Assignment07.jl</code>.</p></li><li><p>Commit and push your changes.</p></li><li><p>Back on github, click on the <code>Settings</code> tab for your repository, then on <code>Manage access</code>, and add Kevin (<code>@kescobo</code>) as a collaborator.</p></li></ol><h2 id="Building-an-analysis-repository"><a class="docs-heading-anchor" href="#Building-an-analysis-repository">Building an analysis repository</a><a id="Building-an-analysis-repository-1"></a><a class="docs-heading-anchor-permalink" href="#Building-an-analysis-repository" title="Permalink"></a></h2><p>Now, we&#39;ll create a new project to handle our analyses of SARS-CoV2 genomes. While it is possible to keep everything in the same directory, it&#39;s a good idea to keep code that may be useful in many projects (such as your bioinformatics functions) separate from code that is used for one specific purpose.</p><ol><li>Open a terminal, and create a new directory with <code>mkdir</code> for CoV2 analysis (you can put it in the same directory as your other assignments, but don&#39;t put it <em>inside</em> any of those assignments). Use <code>cd</code> to make this new directory your working directory, then enter <code>git init</code> to make this into a git repository.</li><li>On github, create a new repository, copy its url, and use <code>git remote add</code> to set this as <code>origin</code> (refer <a href="../../Lessons/Lesson1/#git-version-control">back to lesson 1</a> if you need a reminder about how these commands work). Don&#39;t forget to add Kevin as a collaborator.</li><li>Create a directory inside this repo called <code>notebooks</code>, and another called <code>data</code>.</li><li>Open the analysis directory in VS Code. Create a file called <code>README.md</code>, give it a title (use markdown to make the title Heading 1 format), and add <code>By &lt;your name&gt;</code> under the title in bold.</li><li>Commit and push your changes. Take a look at the repo on github - your README should show up, correctly formatted, but the directories won&#39;t show up until there are committed files in them.</li><li>Add a file called <code>data.md</code> to the <code>data/</code> directory - This is where we&#39;ll describe how to get the data that we&#39;re going to analyze, but you can leave it blank for now</li><li>Add a file called <code>1-sequence-stats.md</code> to the <code>notebooks</code> directory. Also ok to leave this file blank for now.</li><li>Commit and push your changes.</li></ol><p>At this point, you should have a git repo with the following structure:</p><pre><code class="language-none">./MyAnalysis
├── data
│   └── data.md
├── notebooks
│   └── 1-kmer-counting.md
└── README.md</code></pre><h2 id="Getting-the-data"><a class="docs-heading-anchor" href="#Getting-the-data">Getting the data</a><a id="Getting-the-data-1"></a><a class="docs-heading-anchor-permalink" href="#Getting-the-data" title="Permalink"></a></h2><p>An enormous amount of data has already been generated on SARS-CoV2 (the virus) and COVID-19 (the disease caused by CoV2). You can get links to much of it <a href="https://www.ncbi.nlm.nih.gov/sars-cov-2/">on NCBI</a> - the National Center for Biotechnology Information.</p><p>We&#39;re particularly interested in the genome sequences of SARS-related corona viruses. Go to the <a href="https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus?SeqType_s=Nucleotide&amp;VirusLineage_ss=SARS-CoV-2,%20taxid:2697049">Virus search page</a>, and use the search function to see all of the genomes of SARS-CoV (Severe accute respiratory syndrom-related Coronavirus), MERS-CoV (Middle East repiratory syndrom-related Coronavirus), and SARS-CoV2 (the virus that causes COVID19).</p><p>Refine the search a bit by selecting only complete genomes (under &quot;Nucleotide completeness&quot;). At the time I write this, there are just under 7300 such records, though there may be more by the time you search.</p><p>The search parameters you enter are stored in the url. Copy the url and put it into your <code>data/data.md</code> file, explaining what it is. It&#39;s always a good idea to keep track of where data comes from so that your work is reproducible by someone else.</p><p>Click &quot;download&quot;, make sure &quot;FASTA Nucleotide&quot; is selected, and download all records. When it asks about the FASTA definition line (the header), build a custom header and be sure to include the Accession number (this is a unique identifier), the species, and the Geo location. </p><p>Download the sequences, and put them into your <code>data/</code> directory with a more descriptive name than <code>sequences.fasta</code>. Be sure to describe what you did in <code>data/data.md</code>.</p><div class="admonition is-success"><header class="admonition-header">Tip</header><div class="admonition-body"></div></div><p>Usually, it&#39;s not a good idea to add and commit large files    like genome sequences to the git repository,    especially if those files are likely to change often    (since <code>git</code> stores a record of all of the different versions).</p><p>In this case, our file is not going to change, and it&#39;s not <strong>that</strong> big,    so it&#39;s ok to commit it, but if you want to avoid commiting it,    and to make <code>git</code> forget about trying to track it,    you can create a file called <code>.gitignore</code> inside your repo,    and put <code>data/&lt;filename&gt;</code> in it.    Be sure to add and commit the <code>.gitignore</code> file!</p><h2 id="Setting-up-julia-project"><a class="docs-heading-anchor" href="#Setting-up-julia-project">Setting up julia project</a><a id="Setting-up-julia-project-1"></a><a class="docs-heading-anchor-permalink" href="#Setting-up-julia-project" title="Permalink"></a></h2><p>Open a julia REPL in your analysis directory (you can do this from VS Code or by opening a new REPL and using <code>cd()</code>).</p><p>Now, activate the environment from the Pkg REPL, and use <code>dev</code> (short for <code>develop</code>) to install your bioinformatics project. To do this, you can use the path (relative or absolute) to the bioinformatics project directory.</p><p>For example, if it is in your <code>Documents/</code> directory, you might enter</p><pre><code class="language-none">pkg&gt; dev /Users/&lt;username&gt;/Documents/BioinformaticsBISC195</code></pre><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"></div></div><p>As with <code>$</code> in shell prompts, the <code>pkg&gt;</code> is an indicator    that you should be using the Pkg REPL (accessible by pressing <code>]</code>),    and isn&#39;t part of the actual command.</p><p>Once you&#39;ve done this, you will be able to do <code>using BioinformaticsBISC195</code> in your julia code, as long as this julia project is active.</p><p>In addition, use <code>add</code> to install the packages <code>Revise</code>, <code>Plots</code>, and <code>DataFrames</code>.  We&#39;ll use those a bit more later.</p><p>Add and commit the <code>Project.toml</code> and <code>Manifest.toml</code> files,</p><h2 id="Using-the-notebook"><a class="docs-heading-anchor" href="#Using-the-notebook">Using the notebook</a><a id="Using-the-notebook-1"></a><a class="docs-heading-anchor-permalink" href="#Using-the-notebook" title="Permalink"></a></h2><p>In the <code>notebooks/1-sequence-stats.md</code> file, write julia code to calculate the mean and standard deviation of the lengths and gc content of your coronavirus genomes, as you did in <a href="../Assignment06/#assignment06">Assignment 6</a>. Include a title (with header format) and a description of what your code is doing. </p><p>Remember, you can use the code you&#39;ve written in your bioinformatics package just by doing <code>using &lt;PackageName&gt;</code>. Also remember, in markdown files, julia code should go in &quot;code fences.&quot; Eg.</p><pre><code class="language-none"># Some Coronavirus Analysis

Here&#39;s some description of what I&#39;m doing.
I downloaded coronavirus geneomes from NCBI
(see `data/data.md` for more details)

Etc.

```julia
using BioinformaticsBISC195

genomes = parse_fasta(&quot;../data/cov2_sequences.fasta&quot;)
```

Now I have the genomes, and need to do stats.
Note - your descriptions should be... more descriptive.

```julia
using Statistics
#more code...

```

etc..</code></pre><p>If you&#39;re using VS Code to edit your markdown, and have both the <code>julia</code> and <code>markdown julia</code> extensions installed, you can execute code within code fences in exactly the same way as you&#39;ve been doing in <code>.jl</code> files.</p><h2 id="No-tests"><a class="docs-heading-anchor" href="#No-tests">No tests</a><a id="No-tests-1"></a><a class="docs-heading-anchor-permalink" href="#No-tests" title="Permalink"></a></h2><p>There are no automated tests for your analysis code - this is research, I don&#39;t know what the right answer is!</p><p>But this means that it&#39;s extra important that you add extra checks to make sure the code is doing what you intend.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../Lessons/Lesson8/">« Lesson docs</a><a class="docs-footer-nextpage" href="../Assignments/">Assignment Index »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Sunday 12 July 2020 00:56">Sunday 12 July 2020</span>. Using Julia version 1.4.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
